sendrecv/gst-rust/src/main.rs

```diff
@@ -0,0 +1,426 @@
+extern crate glib;
+extern crate gstreamer as gst;
+extern crate gstreamer_sdp;
+extern crate gstreamer_sdp_sys;
+extern crate gstreamer_webrtc;
+extern crate rand;
+extern crate ws;
+#[macro_use]
+extern crate serde_json;
+
+use glib::translate::*;
+use gst::prelude::*;
+use gst::{BinExt, ElementExt};
+use rand::Rng;
+use std::sync::atomic::{AtomicUsize, Ordering};
+use std::sync::Arc;
+
+#[derive(PartialEq, Eq, Debug)]
+enum AppState {
+    // AppStateUnknown = 0,
+    AppStateErr = 1,
+    ServerConnecting = 1000,
+    ServerConnectionError,
+    ServerConnected,
+    ServerRegistering = 2000,
+    ServerRegisteringError,
+    ServerRegistered,
+    ServerClosed,
+    PeerConnecting = 3000,
+    PeerConnectionError,
+    PeerConnected,
+    PeerCallNegotiating = 4000,
+    PeerCallStarted,
+    PeerCallStopping,
+    PeerCallStopped,
+    PeerCallError,
+    None,
+}
+
+fn check_plugins() -> bool {
+    let needed = vec![
+        "opus",
+        "vpx",
+        "nice",
+        "webrtc",
+        "dtls",
+        "srtp",
+        "rtpmanager",
+        "videotestsrc",
+        "audiotestsrc",
+    ];
+    let registry = gst::Registry::get();
+    let mut ret = true;
+    for plugin_name in needed {
+        let plugin = registry.find_plugin(&plugin_name.to_string());
+        if plugin.is_none() {
+            println!("Required gstreamer plugin '{}' not found", plugin_name);
+            ret = false;
+        }
+    }
+    return ret;
+}
+
+fn setup_call(app_state: &Arc<AtomicUsize>, out: &ws::Sender) -> AppState {
+    app_state.store(AppState::PeerConnecting as usize, Ordering::Relaxed);
+    println!("Setting up signalling server call with 1");
+    out.send("SESSION 1").unwrap();
+    return AppState::PeerConnecting;
+}
+
+fn register_with_server(app_state: &Arc<AtomicUsize>, out: &ws::Sender) -> AppState {
+    app_state.store(AppState::ServerRegistering as usize, Ordering::Relaxed);
+    let our_id = rand::thread_rng().gen_range(10, 10_000);
+    println!("Registering id {} with server", our_id);
+    out.send(format!("HELLO {}", our_id)).unwrap();
+    return AppState::ServerRegistering;
+}
+
+fn sdp_message_as_text(offer: gstreamer_webrtc::WebRTCSessionDescription) -> Option<String> {
+    unsafe {
```

Comment:
https://github.com/sdroege/gstreamer-rs/issues/108#issuecomment-393505826 for why this is necessary currently. We should fix that before merging this PR :)

Reply:
I've added those now here https://github.com/sdroege/gstreamer-rs/commit/85ca3b94a34d0236314aac900d5926c81ba73d99

Reply:
Lovely timing, as mine was incorrect: https://github.com/maxmcd/gstreamer-rs/commit/dd0918b05d9cd70c0aa359762080545e1cb1000e

:)

Reply:
A mutable borrow is not needed here as all these are immutable fields after all, and `from_glib_none` is needed here because we don't take ownership of the field (otherwise we would steal it from the struct). Ideally we would just borrow it, but there's no borrowing mechanism in `glib-rs` at this point that could be used for this (and generally borrowing from GLib types is dangerous by default except for some special cases), so we have to copy here.

Reply:
Let me know if my version works and if there are any other problems :) And thanks again for working on this

Reply:
Your version works perfectly. Replaced the unsafe call.